src/statd/journal.c

```diff
@@ -0,0 +1,198 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <sysrepo.h>
+#include <ev.h>
+#include <string.h>
+#include <errno.h>
+#include <time.h>
+#include <sys/stat.h>
+#include <pthread.h>
+#include <dirent.h>
+
+#include <srx/common.h>
+
+#include "journal.h"
+
+#define JOURNAL_DIR "/var/lib/statd"
+#define DUMP_FILE "/var/lib/statd/operational.json"
+#define DUMP_INTERVAL 300.0  /* 5 minutes in seconds */
+
+static void journal_stop_cb(struct ev_loop *loop, struct ev_async *, int)
+{
+	DEBUG("Journal thread stop signal received");
+	ev_break(loop, EVBREAK_ALL);
+}
+
+static void get_timestamp_filename(char *buf, size_t len, time_t ts)
+{
+	struct tm *tm = gmtime(&ts);
+
+	snprintf(buf, len, "%04d%02d%02d-%02d%02d%02d.json",
+		 tm->tm_year + 1900, tm->tm_mon + 1, tm->tm_mday,
+		 tm->tm_hour, tm->tm_min, tm->tm_sec);
+}
+
+/* Create timestamped snapshot and update operational.json symlink */
+static int create_snapshot(const struct lyd_node *tree)
+{
+	char timestamp_file[300];
+	char timestamp_path[512];
+	time_t now;
+	int ret;
+
+	now = time(NULL);
+	get_timestamp_filename(timestamp_file, sizeof(timestamp_file), now);
+	snprintf(timestamp_path, sizeof(timestamp_path), "%s/%s",
+		 JOURNAL_DIR, timestamp_file);
+
+	/* Write timestamped snapshot directly */
+	ret = lyd_print_path(timestamp_path, tree, LYD_JSON,
+			     LYD_PRINT_WITHSIBLINGS);
```

Comment:
Did some quick testing on a 28-port system with more or less an empty config:

```
root@test-05-25-f8:/tmp$ sysrepocfg -X -d operational -f json >oper.json
root@test-05-25-f8:/tmp$ ll -h oper.json
-rw-r--r--    1 root     root      237.9K Dec  9 22:36 oper.json
root@test-05-25-f8:/tmp$ gzip oper.json
root@test-05-25-f8:/tmp$ ll -h oper.json.gz
-rw-r--r--    1 root     root       19.6K Dec  9 22:37 oper.json.gz
```
Should we pass this through `libz` before going to disk? I'd hate for this storage to become an issue on smaller systems or as the amount of generated data grows. With compression, disk usage after one year should go from 14MB down to 1MB with today's feature-set. That should set us up to not have to think about this for a _long_ time.